packages/cli/src/commands/rewards/slasher.ts

```diff
@@ -0,0 +1,269 @@
+
+    const done = () =>
+      (endBlock && blockNumber >= endBlock) ||
+      (maxSlashAttempts && slashAttempts >= maxSlashAttempts)
+    while (!done()) {
+      // Don't use web3.eth.subscribe in case we're connected over HTTP
+      const newBlockNumber = await this.kit.web3.eth.getBlockNumber()
+      while (blockNumber < newBlockNumber && !done()) {
+        blockNumber++
+
+        const newEpochNumber = await this.kit.getEpochNumberOfBlock(blockNumber)
+        if (epochNumber !== newEpochNumber) {
+          epochNumber = newEpochNumber
+          console.info(`New epoch: ${epochNumber}`)
+
+          // Map validator indices across epochs, initializing validatorDownSince to -1
+          const oldEpochSigners = await election.getValidatorSigners(blockNumber - 1)
+          const newEpochSigners = await election.getValidatorSigners(blockNumber)
+          validatorSigners = newEpochSigners
+          validatorSet = await concurrentMap(10, validatorSigners, (x: Address) =>
+            validators.getValidatorFromSigner(x)
+          )
+          validatorDownSince = await mapAddressListDataOnto(
+            validatorDownSince,
+            oldEpochSigners,
+            newEpochSigners,
+            -1
+          )
+        }
+
+        const block = await this.kit.web3.eth.getBlock(blockNumber)
+        console.info(`New block: ${blockNumber}`)
+
+        const istanbulExtra = parseBlockExtraData(block.extraData)
+        const seal = istanbulExtra.parentAggregatedSeal
+        const sealBlockNumber = blockNumber - 1
+
+        // Check for each validator in tip parentSeal
+        for (let i = 0; i < validatorSet.length; i++) {
+          const validatorUp = bitIsSet(seal.bitmap, i)
+          if (validatorUp) {
+            if (validatorDownSince[i] >= 0) {
+              const downtime = sealBlockNumber - validatorDownSince[i]
+              validatorDownSince[i] = -1
+              const validator = validatorSet[i]
+              console.info(
+                `Validator:${i}: "${validator.name}" back at block ${sealBlockNumber}, downtime=${downtime}`
+              )
+            }
+          } else {
+            if (validatorDownSince[i] < 0) {
+              validatorDownSince[i] = sealBlockNumber
+              const validator = validatorSet[i]
+              console.info(
+                `Validator:${i}: "${validator.name}" missing from block ${sealBlockNumber}`
+              )
+            } else if (
+              sealBlockNumber - validatorDownSince[i] + 1 >=
+              slashableDowntime + extraWait
+            ) {
+              const downtime = sealBlockNumber - validatorDownSince[i] + 1
+              const downtimeEndBlock = sealBlockNumber - extraWait
+              const validator = validatorSet[i]
+              const signer = validatorSigners[i]
+              const alreadySlashedSigner = alreadySlashed[normalizeAddress(signer)]
```

Comment:
Only check if alreadySlashedSigner if startBlock has been supplied, and explain why we don't need to check if we're caught up to tip.